File: game/tv_show/in_cave.py
from cave.cave_game import CaveGame
from game_data import GameData
from phone_events import PhoneEvents
from state import State
from tv_show.attract import Attract
from tv_show.ending import Ending


class InCave(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.cave = CaveGame(context)

    def process_events(self, phone_events: PhoneEvents):
        if phone_events.hungup:
            return Attract

        if self.cave.ended:
            return Ending

        return self.cave.process_events(phone_events)

    def render(self, screen):
        self.cave.render(screen)

    def on_exit(self):
        super().on_exit()
        self.cave = None

File: game-web/src/tvShow/inCave.ts
/** Port of game/tv_show/in_cave.py */

import { State, type StateClass } from '../state';
import { CaveGame } from '../cave/caveGame';
import { Attract } from './attract';
import { Ending } from './ending';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class InCave extends State {
  private cave: CaveGame | null;

  constructor(context: GameData) {
    super(context);
    this.cave = new CaveGame(context);
  }

  processEvents(events: PhoneEvents): StateClass | null {
    if (events.hungup) return Attract;
    if (this.cave!.ended) return Ending;
    this.cave!.processEvents(events);
    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    this.cave!.render(ctx);
  }

  onExit(): void {
    super.onExit();
    this.cave = null;
  }
}
